src/main/cpp/src/map.cu

```diff
@@ -0,0 +1,92 @@
+#include <cudf/sorting.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/resource_ref.hpp>
+
+namespace spark_rapids_jni {
+
+std::unique_ptr<cudf::column> sort_map_column(cudf::column_view const& map_column,
+                                              cudf::order sort_order,
+                                              rmm::cuda_stream_view stream,
+                                              rmm::device_async_resource_ref mr)
+{
+  auto const& lists_of_structs = cudf::lists_column_view(map_column);
+  auto const structs           = lists_of_structs.child();
+  CUDF_EXPECTS(structs.type().id() == cudf::type_id::STRUCT,
+               "maps_column_view input must have exactly 1 child (STRUCT) column.");
+  CUDF_EXPECTS(structs.num_children() == 2,
+               "maps_column_view key-value struct must have exactly 2 children.");
+  auto keys   = structs.child(0);
+  auto values = structs.child(1);
+  CUDF_EXPECTS(structs.null_count() == 0, "maps_column_view key-value struct must have no null.");
+  CUDF_EXPECTS(keys.null_count() == 0, "maps_column_view keys must have no null.");
+  auto segments = lists_of_structs.offsets();
+
+  auto sorted = cudf::segmented_sort_by_key(cudf::table_view{{keys, values}},
+                                            cudf::table_view{{keys}},
+                                            segments,
+                                            {sort_order},
+                                            {},  // Map keys MUST be not null
+                                            stream,
+                                            mr);
+  stream.synchronize();
+  std::vector<std::unique_ptr<cudf::column>> k_v = sorted->release();
+
+  // Note: The keys in a map MUST not be null, so the struct<Key, Value> MUST not be null too.
+  // Of course, the null count is zero for struct<Key Value> column
+  auto mask = cudf::create_null_mask(keys.size(), cudf::mask_state::UNALLOCATED, stream, mr);
+  stream.synchronize();
+  auto sorted_struct =
+    cudf::make_structs_column(keys.size(), std::move(k_v), 0, std::move(mask), stream, mr);
+
+  // clone segments
+  auto copied_segements = cudf::make_numeric_column(cudf::data_type(segments.type().id()),
+                                                    segments.size(),
+                                                    cudf::mask_state::UNALLOCATED,
+                                                    stream,
+                                                    mr);
+  stream.synchronize();
+  CUDF_CUDA_TRY(cudaMemcpyAsync(copied_segements->mutable_view().data<int32_t>(),
+                                segments.data<int32_t>(),
+                                segments.size() * sizeof(int32_t),
+                                cudaMemcpyDeviceToDevice,
+                                stream.value()));
+  stream.synchronize();
```

Comment:
This is just:
```
auto output_segments = std::make_unique<cudf::column>(segments);
```

Reply:
IIUC: The passed in map ColumnView is managed by Java side ColumnVector via  refCount in it. If the refCount is zero, the Java side will release the whole map column including the `offsets` sub column.  If we use `offsets` directly, we may reference released memory.

Reply:
No we are not using `offsets` in the final output. We copy it by the constructor of `cudf::column` as my code above.